feat: Add create_ignore_already_exists to workload_identity

flozzone opened the pull request:
Adds [create_ignore_already_exists](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/google_service_account#create_ignore_already_exists) to workload-identity module to ignore an already existing google-service-account when hitting the [Provider produced inconsistent result after apply](https://github.com/hashicorp/terraform-provider-google/issues/18087) issue.

Fixes #2141

apeabody commented:
/gcbrun

apeabody reviewed (changes requested):
Thanks for the contribution @flozzone!


From the linter:
```
Running terraform fmt
modules/workload-identity/main.tf
--- old/modules/workload-identity/main.tf
+++ new/modules/workload-identity/main.tf
@@ -42,11 +42,11 @@
 resource "google_service_account" "cluster_service_account" {
   count = var.use_existing_gcp_sa ? 0 : 1
 
-  account_id                    = local.gcp_given_name
-  display_name                  = coalesce(var.gcp_sa_display_name, substr("GCP SA bound to K8S SA ${local.k8s_sa_project_id}[${local.k8s_given_name}]", 0, 100))
-  description                   = var.gcp_sa_description
-  project                       = var.project_id
-  create_ignore_already_exists  = var.create_ignore_already_exists
+  account_id                   = local.gcp_given_name
+  display_name                 = coalesce(var.gcp_sa_display_name, substr("GCP SA bound to K8S SA ${local.k8s_sa_project_id}[${local.k8s_given_name}]", 0, 100))
+  description                  = var.gcp_sa_description
+  project                      = var.project_id
+  create_ignore_already_exists = var.create_ignore_already_exists
 }
 
 resource "kubernetes_service_account" "main" {
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
```

apeabody commented:
/gcbrun

flozzone commented:
Hi @apeabody, I've corrected linting errors but now the build is failing and I cannot see any logs at https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/2142/checks?check_run_id=31930708825 and I rebased now to see if that caused the issue.

apeabody commented:
/gcbrun

apeabody commented:
> Hi @apeabody, I've corrected linting errors but now the build is failing and I cannot see any logs at https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/2142/checks?check_run_id=31930708825 and I rebased now to see if that caused the issue.

I've triggered the test, but we've been having unrelated intermittent test failures (timeout) that we are also working to resolve.

apeabody commented:
/gcbrun

apeabody commented:
/gcbrun

apeabody commented:
/gcbrun

apeabody commented on modules/workload-identity/variables.tf:
```diff
@@ -137,3 +137,9 @@ variable "gcp_sa_description" {
+variable "create_ignore_already_exists" {
+  description = "If set to true, skip service account creation if a service account with the same email already exists."
+  type        = bool
+  default     = false
```
I wonder if `null` would be a option for default so there is no change to the existing `google_service_account`?

apeabody commented on modules/workload-identity/variables.tf:
```diff
@@ -137,3 +137,9 @@ variable "gcp_sa_description" {
   }
 }
+
+variable "create_ignore_already_exists" {
```
Could we add `gcp_sa_` to the variable name similar to the existing variables?

flozzone replied:
Totally forgot to look for naming conventions :+1:

apeabody reviewed (changes requested):
Thanks for the contribution @flozzone!

A few small notes before we merge.

apeabody commented:
/gcbrun

inverse-panda commented:
I thought the workflow is to use use_existing_gcp_sa to basically use an exising google service account.
Now I got the this error in terraform cloud
![image](https://github.com/user-attachments/assets/92db1c99-ee05-4e4e-adad-085951951472)

Is there something I need to do as a user of this module? Below you can see my terraform object
```
module "dataray_mlflow_stage_indentity" {
  source       = "terraform-google-modules/kubernetes-engine/google//modules/workload-identity"
  name         = "dataray-mlflow-stage"
  namespace    = "stage"
  cluster_name = var.cluster_name
  location     = var.zone
  project_id   = var.project
  roles        = ["roles/storage.admin", "roles/storage.objectCreator", "roles/cloudsql.client", "roles/bigquery.dataOwner", "roles/bigquery.user"]
}
```

apeabody commented:
> I thought the workflow is to use use_existing_gcp_sa to basically use an exising google service account. Now I got the this error in terraform cloud ![image](https://private-user-images.githubusercontent.com/128373405/381895477-92db1c99-ee05-4e4e-adad-085951951472.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAzOTIzMzIsIm5iZiI6MTczMDM5MjAzMiwicGF0aCI6Ii8xMjgzNzM0MDUvMzgxODk1NDc3LTkyZGIxYzk5LWVlMDUtNGU0ZS1hZGFkLTA4NTk1MTk1MTQ3Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDMxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAzMVQxNjI3MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMzAzMGU2ZWM4YzhhNzQzOWE4YTkzMTFjNGI3YzJlM2ZiN2I5OGI0NDFmOWI3YTMwNGVlOWFmOTczNWJkYWVmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.5ynyNc2qs2dVq6lufEHYTfaq0bHlLrzOJ7JqfNGIODA)
> 
> Is there something I need to do as a user of this module? Below you can see my terraform object
> 
> ```
> module "dataray_mlflow_stage_indentity" {
>   source       = "terraform-google-modules/kubernetes-engine/google//modules/workload-identity"
>   name         = "dataray-mlflow-stage"
>   namespace    = "stage"
>   cluster_name = var.cluster_name
>   location     = var.zone
>   project_id   = var.project
>   roles        = ["roles/storage.admin", "roles/storage.objectCreator", "roles/cloudsql.client", "roles/bigquery.dataOwner", "roles/bigquery.user"]
> }
> ```

Hi @inverse-panda - Can you share the version of the provider (`terraform version`) and module you are using?  The output of `terraform plan` mightal also be useful.

I would also strongly recommend specifying the major version of the module such as:
```
module "workload_identity_existing_ksa" {
  source  = "terraform-google-modules/kubernetes-engine/google//modules/workload-identity"
  version = "~> 34.0"
```

flozzone commented:
@inverse-panda I get the point, it might be a bit misleading now. But gcp_sa_create_ignore_already_exists is used in case Terraform fails to create the SA due to GCP IAM eventual consistency principle. See https://github.com/hashicorp/terraform-provider-google/issues/18087

And today I found out that recent versions of the Google Provider has fixed that problem on provider end https://github.com/hashicorp/terraform-provider-google/releases/tag/v6.7.0 with https://github.com/hashicorp/terraform-provider-google/pull/19727

flozzone commented:
@apeabody It seems I should have checked provider version compatibility and bump the required version

apeabody commented:
> @apeabody It seems I should have checked provider version compatibility and bump the required version

Hi @flozzone v34 was specifically released to still support TPGv5.  The upcoming V35 will be TPGv6 only and likely require at least TPG v6.7+.  Do you think we should revert this addition in v35?

flozzone commented:
Hi @apeabody, I see `create_ignore_already_exists` has been introduced in [TPGv5.12](https://github.com/hashicorp/terraform-provider-google/releases/tag/v5.12.0) with [#16927](https://github.com/hashicorp/terraform-provider-google/pull/16927), so TPGv5 already supports it. So I think @inverse-panda issue is caused by TPG minimum version constraint of `>= 3.39.0` in [required_providers](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/modules/workload-identity/versions.tf#L24). If we would bump it to `>= v5.12.0`, it would still be TPGv5 compatible.

So people having the [Provider produced inconsistent result after apply](https://github.com/hashicorp/terraform-provider-google/issues/18087) when creating service accounts would either have to use TPGv6.7 for a proper fix or use `create_ignore_already_exists` workaround from TPGv5.12.

apeabody commented:
> Hi @apeabody, I see `create_ignore_already_exists` has been introduced in [TPGv5.12](https://github.com/hashicorp/terraform-provider-google/releases/tag/v5.12.0) with [#16927](https://github.com/hashicorp/terraform-provider-google/pull/16927), so TPGv5 already supports it. So I think @inverse-panda issue is caused by TPG minimum version constraint of `>= 3.39.0` in [required_providers](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/modules/workload-identity/versions.tf#L24). If we would bump it to `>= v5.12.0`, it would still be TPGv5 compatible.
> 
> So people having the [Provider produced inconsistent result after apply](https://github.com/hashicorp/terraform-provider-google/issues/18087) when creating service accounts would either have to use TPGv6.7 for a proper fix or use `create_ignore_already_exists` workaround from TPGv5.12.

Thanks @flozzone! Would you be able to open a PR bumping the version to `>= v5.12.0`?

flozzone commented:
Here we go #2170 @apeabody

gustysap commented:
Hi, I use TPG ~> 4.0 and got this

```
│   on .terraform/modules/wi-name/modules/workload-identity/main.tf line 49, in resource "google_service_account" "cluster_service_account":
│   49:   create_ignore_already_exists = var.gcp_sa_create_ignore_already_exists
```

I need to upgrade the TPG or what ? currently I manually remove the 

`create_ignore_already_exists = var.gcp_sa_create_ignore_already_exists
`
in the main.tf of module

apeabody commented:
> Hi, I use TPG ~> 4.0 and got this
> 
> ```
> │   on .terraform/modules/wi-name/modules/workload-identity/main.tf line 49, in resource "google_service_account" "cluster_service_account":
> │   49:   create_ignore_already_exists = var.gcp_sa_create_ignore_already_exists
> ```
> 
> I need to upgrade the TPG or what ? currently I manually remove the
> 
> `create_ignore_already_exists = var.gcp_sa_create_ignore_already_exists ` in the main.tf of module

Hi @gustysap In v34 the `workload-identity` sub-module actually requires TPG >-=5.12.0.  You may be able to use `terraform init -upgrade`.

inverse-panda commented:
Thanks a lot @flozzone, updating the version to  `TGP = "~> 5.12.0"` solved my issue.
